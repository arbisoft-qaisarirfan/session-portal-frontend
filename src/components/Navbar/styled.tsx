import InputBase, { inputBaseClasses } from "@mui/material/InputBase";
import { alpha, css, styled } from "@mui/material/styles";
import { typographyClasses } from "@mui/material/Typography";

export const Logo = styled("div", { name: "Logo" })(({ theme }) => {
  return css`
    align-items: center;
    display: flex;
    width: 240px;

    svg {
      color: red;
      margin-right: 8px;
    }

    .${typographyClasses.h6} {
      color: ${theme.palette.common.white};
      font-size: 16px;
      font-style: normal;
      font-weight: 700;
      line-height: normal;
      text-decoration: none;
    }
  `;
});

export const Search = styled("div", { name: "Search" })(({ theme }) => {
  return css`
    align-items: center;
    background-color: ${alpha(theme.palette.common.white, 0.15)};
    border-radius: ${theme.shape.borderRadius * 2}px;
    display: flex;
    justify-content: space-between;
    margin-left: 0;
    width: 450px;

    &:hover {
      background-color: ${alpha(theme.palette.common.white, 0.25)};
    }
  `;
});

export const SearchIconWrapper = styled("div", { name: "SearchIconWrapper" })(({ theme }) => {
  return css`
    align-items: center;
    background-color: ${alpha(theme.palette.common.white, 0.25)};
    border-bottom-right-radius: ${theme.shape.borderRadius * 2}px;
    border-top-right-radius: ${theme.shape.borderRadius * 2}px;
    display: flex;
    height: 100%;
    justify-content: center;
    padding: ${theme.spacing(0, 2)};
    pointer-events: none;
  `;
});

export const StyledInputBase = styled(InputBase, { name: "StyledInputBase" })(
  ({ theme }) => css`
    color: inherit;
    width: 100%;
    & .${inputBaseClasses.input} {
      padding: ${theme.spacing(1, 1, 1, 1)};
      padding-right: calc(1em + ${theme.spacing(4)});
    }
  `
);
