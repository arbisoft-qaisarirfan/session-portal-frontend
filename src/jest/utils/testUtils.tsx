import React, { ReactElement } from "react";

import "@testing-library/jest-dom";
import { render, RenderOptions } from "@testing-library/react";

import ThemeProvider from "@/components/theme/theme-provider";
import { Providers } from "@/redux/store/provider";

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) =>
  render(ui, {
    wrapper: ({ children }) => (
      <Providers>
        <ThemeProvider>{children}</ThemeProvider>
      </Providers>
    ),
    ...options,
  });

export * from "@testing-library/react";

export { customRender };
