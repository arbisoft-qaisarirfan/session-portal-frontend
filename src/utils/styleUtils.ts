export const shouldForwardProp = (propName: string) => !propName.startsWith("$");
