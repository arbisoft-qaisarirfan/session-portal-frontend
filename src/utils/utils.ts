export function secondsToTime(seconds: number): string {
  const hours = Math.floor(seconds / 3600);
  const minutes = Math.floor((seconds % 3600) / 60);

  const formatNumber = (num: number): string =>
    num.toLocaleString(undefined, {
      useGrouping: false,
      minimumIntegerDigits: 2,
    });

  return `${formatNumber(hours)}:${formatNumber(minutes)}`;
}
