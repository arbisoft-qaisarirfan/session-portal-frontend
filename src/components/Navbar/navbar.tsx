import { useState } from "react";

import SearchIcon from "@mui/icons-material/Search";
import YouTube from "@mui/icons-material/YouTube";
import AppBar from "@mui/material/AppBar";
import Avatar from "@mui/material/Avatar";
import Box from "@mui/material/Box";
import Container from "@mui/material/Container";
import IconButton from "@mui/material/IconButton";
import Menu from "@mui/material/Menu";
import MenuItem from "@mui/material/MenuItem";
import Toolbar from "@mui/material/Toolbar";
import Tooltip from "@mui/material/Tooltip";
import Typography from "@mui/material/Typography";
import { useDispatch, useSelector } from "react-redux";

import { selectUserInfo } from "@/redux/login/selectors";
import { loginActions } from "@/redux/login/slice";
import { persistor } from "@/redux/store/configureStore";

import ThemeToggle from "../ThemeToggle";
import { Logo, Search, SearchIconWrapper, StyledInputBase } from "./styled";

const settings = ["Profile", "Account", "Dashboard"];

function Navbar() {
  const dispatch = useDispatch();
  const userInfo = useSelector(selectUserInfo);

  const [searchQuery, setSearchQuery] = useState("");
  const [anchorElUser, setAnchorElUser] = useState<null | HTMLElement>(null);

  const handleOpenUserMenu = (event: React.MouseEvent<HTMLElement>) => {
    setAnchorElUser(event.currentTarget);
  };

  const handleCloseUserMenu = () => {
    setAnchorElUser(null);
  };

  const handleLogout = async () => {
    dispatch(loginActions.logout());
    await persistor.purge();
    persistor.persist();
    handleCloseUserMenu();
  };

  const handleSearchQuery = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
  };

  return (
    <AppBar position="static" data-testid="navbar">
      <Container maxWidth={false}>
        <Toolbar disableGutters>
          <Logo>
            <YouTube />
            <Typography variant="h6" noWrap sx={{ display: { xs: "none", md: "flex" } }}>
              Arbisoft Sessions Portal
            </Typography>
          </Logo>

          <Box sx={{ flexGrow: 1, display: "flex", justifyContent: "center" }}>
            <Search onSubmit={handleSearchQuery}>
              <StyledInputBase
                placeholder="Search..."
                inputProps={{ "aria-label": "search" }}
                value={searchQuery}
                onInput={(e: React.ChangeEvent<HTMLInputElement>) => setSearchQuery(e.target.value)}
              />
              <SearchIconButton type="submit">
                <SearchIcon data-testid="SearchIcon" />
              </SearchIconButton>
            </Search>
          </Box>
          <Box sx={{ flexGrow: 0, width: 240, display: "flex", justifyContent: "flex-end" }}>
            <Tooltip title="Open settings">
              <IconButton data-testid="avatar-btn" onClick={handleOpenUserMenu} sx={{ p: 0 }}>
                <Avatar variant="rounded" alt={userInfo.full_name ?? ""} src={userInfo.avatar ?? ""} />
              </IconButton>
            </Tooltip>
            <Menu
              anchorEl={anchorElUser}
              anchorOrigin={{
                vertical: "bottom",
                horizontal: "right",
              }}
              keepMounted
              transformOrigin={{
                vertical: "top",
                horizontal: "right",
              }}
              open={Boolean(anchorElUser)}
              onClose={handleCloseUserMenu}
            >
              {settings.map((setting) => (
                <MenuItem data-testid={setting} key={setting} onClick={handleCloseUserMenu}>
                  <Typography>{setting}</Typography>
                </MenuItem>
              ))}
              <MenuItem onClick={handleLogout}>
                <Typography>Logout</Typography>
              </MenuItem>
              <MenuItem disableRipple disableTouchRipple>
                <ThemeToggle />
              </MenuItem>
            </Menu>
          </Box>
        </Toolbar>
      </Container>
    </AppBar>
  );
}

export default Navbar;
