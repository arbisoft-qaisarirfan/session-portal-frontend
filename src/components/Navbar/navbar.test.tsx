import { act, fireEvent, customRender as render, screen } from "@/jest/utils/testUtils";

import Navbar from "./navbar";

describe("Navbar Component", () => {
  test("should renders Navbar without crashing", () => {
    render(<Navbar />);
    expect(screen.getByTestId("navbar")).toBeInTheDocument();
  });

  test("should renders the logo with YouTube icon", () => {
    render(<Navbar />);
    expect(screen.getByTestId("navbar")).toBeInTheDocument();
    expect(screen.getByText("Arbisoft Sessions Portal")).toBeInTheDocument();
  });

  test("should renders search input", () => {
    render(<Navbar />);
    expect(screen.getByPlaceholderText("Search...")).toBeInTheDocument();
  });

  test("should renders avatar button", () => {
    render(<Navbar />);
    expect(screen.getByRole("button")).toBeInTheDocument();
  });

  test("should opens user menu when avatar is clicked", () => {
    render(<Navbar />);
    const avatarButton = screen.getByRole("button");
    fireEvent.click(avatarButton);
    expect(screen.getByText("Profile")).toBeInTheDocument();
  });

  test("should closes user menu when menu item is clicked", () => {
    render(<Navbar />);
    const avatarButton = screen.getByRole("button");
    fireEvent.click(avatarButton);
    const menuItem = screen.getByText("Profile");
    fireEvent.click(menuItem);
    expect(menuItem).not.toBeVisible();
  });

  test("should search input is focusable", () => {
    render(<Navbar />);
    const searchInput = screen.getByPlaceholderText("Search...");
    act(() => {
      searchInput.focus();
      expect(searchInput).toHaveFocus();
    });
  });

  test("should user menu contains all settings", () => {
    render(<Navbar />);
    const avatarButton = screen.getByRole("button");
    fireEvent.click(avatarButton);
    const settings = ["Profile", "Account", "Dashboard", "Logout"];
    settings.forEach((setting) => {
      expect(screen.getByText(setting)).toBeInTheDocument();
    });
  });

  test("should applies correct styles for the search box", () => {
    render(<Navbar />);
    const searchBox = screen.getByPlaceholderText("Search...");
    expect(searchBox).toHaveStyle("color: currentColor");
  });

  test("should tooltip is present on avatar button", async () => {
    render(<Navbar />);
    const avatarButton = screen.getByRole("button");
    fireEvent.mouseOver(avatarButton);
    expect(await screen.findByText("Open settings")).toBeInTheDocument();
  });

  test("should search icon is displayed", () => {
    render(<Navbar />);
    expect(screen.getByTestId("SearchIcon")).toBeInTheDocument();
  });

  test("should snapshot test for Navbar", () => {
    const { asFragment } = render(<Navbar />);
    expect(asFragment()).toMatchSnapshot();
  });
});
