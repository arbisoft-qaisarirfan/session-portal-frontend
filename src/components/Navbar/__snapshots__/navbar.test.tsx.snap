// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar Component should match snapshot 1`] = `
<DocumentFragment>
  <header
    class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-ptdwpq-MuiPaper-root-MuiAppBar-root"
    data-testid="navbar"
    style="--Paper-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);"
  >
    <div
      class="MuiContainer-root css-12pwiy3-MuiContainer-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-regular css-1u9o2lj-MuiToolbar-root"
      >
        <div
          class="css-co8q06-Logo-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
            data-testid="YouTubeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 15l5.19-3L10 9v6m11.56-7.83c.13.47.22 1.1.28 1.9.07.8.1 1.49.1 2.09L22 12c0 2.19-.16 3.8-.44 4.83-.25.9-.83 1.48-1.73 1.73-.47.13-1.33.22-2.65.28-1.3.07-2.49.1-3.59.1L12 19c-4.19 0-6.8-.16-7.83-.44-.9-.25-1.48-.83-1.73-1.73-.13-.47-.22-1.1-.28-1.9-.07-.8-.1-1.49-.1-2.09L2 12c0-2.19.16-3.8.44-4.83.25-.9.83-1.48 1.73-1.73.47-.13 1.33-.22 2.65-.28 1.3-.07 2.49-.1 3.59-.1L12 5c4.19 0 6.8.16 7.83.44.9.25 1.48.83 1.73 1.73z"
            />
          </svg>
          <h6
            class="MuiTypography-root MuiTypography-h6 MuiTypography-noWrap css-1xs0nbu-MuiTypography-root"
          >
            Arbisoft Sessions Portal
          </h6>
        </div>
        <div
          class="MuiBox-root css-1rqqp05"
        >
          <form
            class="css-1yzrppa-Search-root"
          >
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary css-ixlbt1-MuiInputBase-root-StyledInputBase-root"
            >
              <input
                aria-label="search"
                class="MuiInputBase-input css-yimnyd-MuiInputBase-input"
                placeholder="Search..."
                type="text"
                value=""
              />
            </div>
            <button
              class="css-140grxj-SearchIconButton-root"
              type="submit"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                />
              </svg>
            </button>
          </form>
        </div>
        <div
          class="MuiBox-root css-1v32olg"
        >
          <button
            aria-label="Open settings"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-1ty4zuc-MuiButtonBase-root-MuiIconButton-root"
            data-mui-internal-clone-element="true"
            data-testid="avatar-btn"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiAvatar-root MuiAvatar-rounded css-14h7kw6-MuiAvatar-root"
            >
              <img
                alt="John Doe"
                class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
                src="https://example.com/avatar.jpg"
              />
            </div>
          </button>
        </div>
      </div>
    </div>
  </header>
</DocumentFragment>
`;
